### PR TITLE
Handle runtime port disconnection and reconnect if needed.

### PR DESCRIPTION
**Pull request by LoneBoco**
This should probably work.  I verified it worked in Edge but I was not able to check Firefox.

### CODE REVIEW

**Comment by MasterOdin** [comment]
Just wanted to say thanks for posting this, it'll probably be a week before I get to this having caught and feeling awful with COVID.

**Comment by LoneBoco** [comment]
I just tested with Firefox and it worked as intended.

File: extension/js/background.js [hunk 1]
```diff
@@ -35,18 +35,33 @@ var salr_client = false;
...
+    let p = chrome.runtime.connect();
+    p.onMessage.addListener(function(data) {
+        if (!salr_client) {
+            salr_client = new SALR(data, chrome.extension.getURL("images/"));
+        }
+    });
+    p.onDisconnect.addListener(() => {
+        port = null;
+        p = null;
+        salr_client = false;
```
Comment (MasterOdin) [inline extension/js/background.js:54]: Can remove this I think. No matter how many times we re setup the port, the instantiated `SALR` class should be essentially the same as neither `data` nor `chrome.extension.getURL("images/")` should change. Additionally, it leverages the `postMessage` function so it doesn't have an explicit binding to the port we make, so even if the definition for `port` changes, it doesn't really matter.

If we did want to reset `SALR` each time to a fresh instance, would have to do `port.postMessage({'message': 'GetPageSettings'})` as we rely on its response to set `data` for `SALR`, and again, that just seems like wasted overhead.

File: extension/js/background.js [hunk 2]
```diff
@@ -35,18 +35,33 @@ var salr_client = false;
...
 port.postMessage({'message': 'GetPageSettings'});
 
+// Create a port to communicate through.
+function createPort() {
+    let p = chrome.runtime.connect();
+    p.onMessage.addListener(function(data) {
+        if (!salr_client) {
+            salr_client = new SALR(data, chrome.extension.getURL("images/"));
+        }
+    });
```
Comment (MasterOdin) [inline extension/js/background.js:50]: Given the other review comment, I'd actually move this listener outside of this function altogether and only bind it to the initial value of `port` as it should live long enough to get the response back from SALR, and that would give slightly less overhead on any additional ports we create. On that vein, I'd actually maybe rework it to be:

```
var port = createPort();
p.onMessage.addListener(function init(data) {
  salr_client = new SALR(data, chrome.extension.getURL("images/"));
  p.onMessage.removeListener(init);
});
```
where once we initialize `salr_client` variable, we can just rid of the listener altogether on the original instance, and also get rid of the `if` statement that was inside of it.

Thoughts?

**Review by MasterOdin — CHANGES_REQUESTED** [review]
Sorry for the FAR too long delay on getting this reviewed! COVID, and then a long sabbatical, and this totally fell off my radar.

I appreciate the PR, and I've just got some minor comments, after which I think should be good to merge. Let me know if my comments make sense.

**Comment by LoneBoco** [comment]
I've made some alterations based on your review.  `salr_client` is only created once, when we get the settings.  I also added some more code comments.

**Review by MasterOdin — APPROVED** [review]

**Comment by MasterOdin** [comment]
Thanks for the PR!

**Pull request merged**